common/dev/peci.c

```diff
@@ -0,0 +1,268 @@
+#include "peci.h"
+
+#include <drivers/peci.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "sensor.h"
+#include "hal_peci.h"
+#include "plat_sensor_table.h"
+
+#define PECI_READ_LEN 5
+#define PECI_RETRY_MAX 6
+
+#ifdef CONFIG_PECI
+static bool peci_get_power(uint8_t sensor_num, int *reading)
+{
+	snr_cfg *cfg = &sensor_config[sensor_config_index_map[sensor_num]];
+	uint8_t cmd = PECI_RD_PKG_CFG0_CMD;
+	uint8_t *read_buf = malloc(2 * PECI_READ_LEN * sizeof(uint8_t));
+	uint8_t index[2] = { 0x03, 0x1F };
+	uint16_t param[2] = { 0x00FF, 0x0000 };
+	uint8_t ret;
+	uint32_t pkg_energy, run_time, diff_energy, diff_time;
+	static uint32_t last_pkg_energy = 0, last_run_time = 0;
+
+	if (read_buf == NULL) {
+		printf("PECI buffer alloc fail\n");
+		return false;
+	}
+
+	uint8_t *complete_code = &read_buf[0];
+	int retry_count = PECI_RETRY_MAX;
+	while (1) {
+		ret = peci_read(cmd, cfg->port, index[0], param[0], PECI_READ_LEN, read_buf);
+		if (ret == 0) {
+			if (*complete_code == PECI_CC_RSP_SUCCESS) {
+				break;
+			}
+		} else if (--retry_count) { // wait for 10ms before next retry
+			k_msleep(10);
+		} else { // all retries failed exit the loop
+			break;
+		}
+	}
+	if (retry_count == 0) {
+		if (ret) {
+			printf("PECI sensor [0x%x] retry read fail, retry time: %d, completion_code: 0x%x\n",
+			       sensor_num, PECI_RETRY_MAX, *complete_code);
+			cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+		} else {
+			switch (*complete_code) {
+			case PECI_CC_RSP_TIMEOUT:
+			case PECI_CC_OUT_OF_RESOURCES_TIMEOUT:
+				printf("PECI sensor [%x] response timeout\n", sensor_num);
+				cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+				break;
+			case PECI_CC_ILLEGAL_REQUEST:
+				printf("PECI sensor [%x] unknown request\n", sensor_num);
+				cfg->cache_status = SENSOR_NOT_FOUND;
+				break;
+			default:
+				printf("PECI sensor [%x] control hardware, firmware or associated logic error\n",
+				       sensor_num);
+				cfg->cache_status = SENSOR_UNSPECIFIED_ERROR;
+				break;
+			}
+		}
+
+		free(read_buf);
```

Comment:
If we're going to use the SAFE_FREE macro we should use it everywhere.

Reply:
The optimization code is based on v2022.07.01. After that, the main branch has been phased in some
commits related to sensor monitoring. Those changes will be phased in main_v2 in the other PRs. This
problem will be solved in those PRs.
Please check #223  .